sys/net/gnrc/network_layer/ipv6/nib/nib.c

```diff
@@ -1027,7 +1027,6 @@ static void _handle_nbr_sol(gnrc_netif_t *netif, const ipv6_hdr_t *ipv6,
                     if (gnrc_netif_is_6lr(netif)) {
                         DEBUG("nib: Storing SL2AO for later handling\n");
                         sl2ao = opt;
-                        break;
```

Comment:
Actually this might be correct: `_copy_and_handle_aro()` calls `_handle_aro()` which in turn calls `_reg_addr_upstream()` which then calls `_handle_sl2ao()` (although I did not check if all the conditions do always apply).

If so better leave a comment here to justify the `break`.

Reply:
If `aro` as an argument of `_copy_and_handle_aro()` is `NULL`, the functions does nothing. Not every NS contains an Address Registration Option. If there is an ARO and `_handle_sl2ao()` is called twice, the second call would determine that there is already a matching Neighbor Cache Entry.

But I wonder if the second call to `_handle_sl2ao()` can be removed.
RFC4861 says:
> The current set of options is defined in such a way that receivers
> can process multiple options in the same packet independently of each
>  other.  In order to maintain these properties, future options SHOULD
>  follow the simple rule:
>    The option MUST NOT depend on the presence or absence of any other
>     options.  The semantics of an option should depend only on the
>     information in the fixed part of the ND packet and on the
>     information contained in the option itself.

Although the 6LowPAN specification did not follow this advise well.
RFC6775:
>  If the source address of the NS is the unspecified address, or if no
>  SLLAO is included, then any included ARO is ignored, that is, the NS
>  is processed as if it did not contain an ARO.

Reply:
Wait! This is a better fix for the problem:
```diff
--- a/sys/net/gnrc/network_layer/ipv6/nib/nib.c
+++ b/sys/net/gnrc/network_layer/ipv6/nib/nib.c
@@ -1027,6 +1027,7 @@ static void _handle_nbr_sol(gnrc_netif_t *netif, const ipv6_hdr_t *ipv6,
                     if (gnrc_netif_is_6lr(netif)) {
                         DEBUG("nib: Storing SL2AO for later handling\n");
                         sl2ao = opt;
+                        break;
                     }
 #endif  /* CONFIG_GNRC_IPV6_NIB_6LR */
                     _handle_sl2ao(netif, ipv6, (const icmpv6_hdr_t *)nbr_sol,
@@ -1044,7 +1045,18 @@ static void _handle_nbr_sol(gnrc_netif_t *netif, const ipv6_hdr_t *ipv6,
                     break;
             }
         }
-        reply_aro = _copy_and_handle_aro(netif, ipv6, nbr_sol, aro, sl2ao);
+        if (aro && sl2ao) {
+        /* If no SLLAO is included, then any included ARO is ignored. */
+            if (!(reply_aro = _copy_and_handle_aro(netif, ipv6, nbr_sol, aro, sl2ao))) {
+            /* If the Length field is not two, or if the Status field is not zero,
+               then the NS is silently ignored.*/
+                return;
+            }
+        }
+        else if (sl2ao) {
+            _handle_sl2ao(netif, ipv6, (const icmpv6_hdr_t *)nbr_sol, sl2ao);
+        }
+
         /* check if target address is anycast */
         if (netif->ipv6.addrs_flags[tgt_idx] & GNRC_NETIF_IPV6_ADDRS_FLAGS_ANYCAST) {
             _send_delayed_nbr_adv(netif, &nbr_sol->tgt, ipv6, reply_aro);
```

If it turns out that the `aro` is of bad format, then we are supposed to ignore the whole NS:
[6.5](https://datatracker.ietf.org/doc/html/rfc6775#section-6.5). Which we did not if we handled the `sllao`.

Reply:
I´m sorry for the sudden change after your approval. I should have given it a second thought earlier.
Does the new fix still make sense to you after a quick look on the section I have cited?

Reply:
If an ACK brings out another critical review that's all for the better :wink: 
I'll try to give this a closer look later.